### ISSUE TITLE
[java] LooseCoupling is at odds with Java 18+ javac linting

### ISSUE THREAD

**Issue by davidburstrom** [pmd/pmd#4360]
**Affects PMD Version:**  

6.53.0 and earlier.

**Rule:**

LooseCoupling: https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#loosecoupling

**Description:**

Since JDK 18+, the linter will [flag fields in serializable classes](https://bugs.openjdk.org/browse/JDK-8274336) if they are declared as types that are non-serializable. In the code sample below, javac will flag the `invalidData` field as suspicious (though contextually, it could reference only serializable implementations), and PMD will flag `validData` as being too tightly coupled.

**Code Sample demonstrating the issue:**

```java
class Class implements Serializable {
  ArrayList<String> validData;
  List<String> invalidData;
}
```

**Expected outcome:**

Ideally, PMD should see that the class is Serializable and relax the requirement on fields to be loosely coupled. Methods shouldn't really be impacted, though.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by jsotuyod** [comment]
Hi, thanks for the report!

I don't think calling this a false positive is accurate 'though… you want the field to be both a `List<String>` and `Serializable`, and even if `ArrayList` fits that bill, there is no reason why other implementations wouldn't (ie: `CopyOnWriteArrayList` is not an `ArrayList`, but is both `Serializable` and a `List`), so the report is actually right.

Unfortunately, due to Java's syntax, the only way to properly express this would require doing something along the lines:

```java
class Foo<T extends List<String> & Serializable> implements Serializable {
  T validData;
}
```

and this is not only more cumbersome, but also pollutes the public API of the class by forcing it to be a generic.

I'm unsure if the rule should be aware of what changes would be required for the fix, or the subjective criteria of when it would or wouldn't be acceptable… maybe this is better off simply being suppressed. Or maybe this is something to handle behind a property… I'm not sure to be honest.

**Comment by davidburstrom** [comment]
Hi!
Well, in the general case I think it makes some sense to have fields that are known to be serializable, if we're talking proper encapsulation and immutability here. My code example in the report is a little contrived, but imagine something like this:

```java
public class MyException extends Exception {
  private ArrayList<String> extraData;

  public MyException(List<String> extraData) {
    this.extraData = new ArrayList<>(extraData);
  }

  public List<String> getExtraData() {
    return Collections.unmodifiableList(extraData);
  }
}
```

Though, as you say, suppression would also do the trick here.

**Comment by oowekyala** [comment]
I think it's reasonable to treat serializable classes specially, because they *need* to be able to enforce that they are recursively serializable. As David mentions the important thing is that the visible API of the class does not couple clients to anything more specific than `List`, so checking only public things is enough (esp method parameters and return types).

Adding a type parameter is not a solution as it is actually forcing all clients to couple to ArrayList or whatever implementation. This is a much bigger problem that just having an ArrayList private field.

**Comment by jsotuyod** [comment]
> the important thing is that the visible API of the class does not couple clients to anything more specific than `List`

I think the rule having properties controlling what it looks at is a great addition. Only checking public API may be a good default, but I can see a case being made for:
* Treating protected and even package-private members as "public API" for classes intended for extension (ie: in PMD we have plenty of such cases).
* Local variables and private members to ensure the cost of change is minimal (even if it's all local, changing the actual type implies multiples changes throughout the file). This is certainly more of a codestyle preference, and I support if this was off by default.

Now, supporting this doesn't require the rule to know about `Serializable` (or any other type). People opting into checking private members should probably suppress these themselves if no better alternative is available, either through a local comment, or more broadly, even an xpath suppression for `isA(java.util.Serializable)`.

**Comment by davidburstrom** [comment]
This is just my opinion, but I'd actually prefer that PMD considers the same entities as today when looking for tight coupling, only making an exception for serializable classes. It's certainly saved me from extensive changes.
